tests/unit/classifier/svm/LibLinear_unittest.cc

```diff
@@ -16,1310 +15,456 @@
-		train_data(0,8)=3;	train_data(1,8)=3;
-		train_data(0,9)=4;	train_data(1,9)=2;
-
-		test_data(0,0)=-1;	test_data(1,0)=1;
-		test_data(0,1)=-1;	test_data(1,1)=0;
-		test_data(0,2)=-1;	test_data(1,2)=-1;
-		test_data(0,3)=-2;	test_data(1,3)=0;
-		test_data(0,4)=0;	test_data(1,4)=-1;
-		test_data(0,5)=2;	test_data(1,5)=2;
-		test_data(0,6)=3;	test_data(1,6)=1;
-		test_data(0,7)=3;	test_data(1,7)=2;
-		test_data(0,8)=3;	test_data(1,8)=3;
-		test_data(0,9)=4;	test_data(1,9)=2;
+		auto pred = ll->apply_binary(test_feats);
+		SG_REF(pred);
 
-		train_feats = new CDenseFeatures<float64_t>(train_data);
-		test_feats = new CDenseFeatures<float64_t>(test_data);
+		auto liblin_accuracy = eval->evaluate(pred, ground_truth);
 
-		SGVector<float64_t> labels(num_samples);
-		for (index_t i = 0; i < num_samples; ++i)
+		if (!biasEnable)
 		{
-			labels[i] = (i < num_samples/2) ? 1.0 : -1.0;
+			for(int i=0;i<t_w.vlen;i++)
+				EXPECT_NEAR(ll->get_w()[i], t_w[i], 1e-5);
 		}
+		else
+		{
+			for(int i=0;i<t_w.vlen;i++)
+				EXPECT_NEAR(ll->get_w()[i], t_w[i], 4e-5);
+			EXPECT_NEAR(ll->get_bias(), t_w[2], 4e-5);
+		}
+		EXPECT_NEAR(liblin_accuracy, 1.0, 1e-5);
 
-		ground_truth = new CBinaryLabels(labels);
+		SG_UNREF(ll);
+		SG_UNREF(eval);
+		SG_UNREF(pred);
+	}
 
-		SG_REF(ground_truth);
+protected:
+	void transpose_feats_matrix()
+	{
+		//Transpose train_feats matrix
+		SGMatrix<float64_t> train_matrix = train_feats->get_feature_matrix();
+		SGMatrix<float64_t>::transpose_matrix(train_matrix.matrix,
+				train_matrix.num_rows, train_matrix.num_cols);
+
+		SG_UNREF(train_feats);
+		train_feats = new CDenseFeatures<float64_t>(train_matrix);
 		SG_REF(train_feats);
-		SG_REF(test_feats);
-}
 
-TEST(LibLinear,train_L2R_LR)
-{
-	LIBLINEAR_SOLVER_TYPE liblinear_solver_type = L2R_LR;
+	}
+	void initialize_train_and_test_and_ground()
+	{
+		index_t num_samples = 50;
+		sg_rand->set_seed(5);
 
-	CDenseFeatures<float64_t>* train_feats = NULL;
-	CDenseFeatures<float64_t>* test_feats = NULL;
-	CBinaryLabels* ground_truth = NULL;
 
-	generate_data_l2(train_feats, test_feats, ground_truth);
+		SGMatrix<float64_t> data =
+			CDataGenerator::generate_gaussians(num_samples, 2, 2);
 
-	CLibLinear* ll = new CLibLinear();
+		CDenseFeatures<float64_t> features(data);
 
-	CBinaryLabels* pred = NULL;
-	float64_t liblin_accuracy;
-	CContingencyTableEvaluation* eval = new CContingencyTableEvaluation();
+		/*
+			Generate the random data
+		*/
+		SGVector<index_t> train_idx(num_samples), test_idx(num_samples);
+		SGVector<float64_t> labels(num_samples);
+		for (index_t i = 0, j = 0; i < data.num_cols; ++i)
+		{
+			if (i % 2 == 0)
+				train_idx[j] = i;
+			else
+				test_idx[j++] = i;
 
-	ll->set_bias_enabled(false);
-	ll->set_features(train_feats);
-	ll->set_labels(ground_truth);
+			labels[i/2] = (i < data.num_cols/2) ? 1.0 : -1.0;
+		}
 
-	ll->set_liblinear_solver_type(liblinear_solver_type);
-	ll->train();
-	pred = ll->apply_binary(test_feats);
-	SG_REF(pred);
+		/*
```

Comment:
minor: could you make those 3 lines one line?

Reply:
Not sure which 3 lines. Do you mean the if statement?

Reply:
the comment ... but it is minor